### PR TITLE
Fix engine.softStart, add tests for engine.brake and ~.softStart

### PR DESCRIPTION
**Pull request by ronso0**
I noticed the argument 'finalRate' passed to engine.softStart() doesn't do what it was intended for:
In case 'finalRate' is not equal to actual deck rate, softStart() just accelerates up to 'finalRate', then jumps to deck rate. I think it's not intuitive to automatically set deck's rate to finalRate (and override tempo slider by that), also I suppose I would to have change a lot in scratchProcess. Therefore I removed 'finalRate'.

[WIP 1]
When brake() is called with factors like 0.5 and below, at a certain point sound is not audible anymore due to the very low frequencies produced by scratchProcess, but deck keeps 'playing' for quite a while before it stops. I'll experiment at which rate playback can be stopped safely.

[WIP 2]
And I'm about to learn how to write tests and want to add some for engine.brake() and engine.softStart(). That's where I need some help:

**brake()** gets arguments `deck`, `activate`, deceleration `factor`, initial `rate`. From this values it sets up a filter, initiates scratchprocess() and sets `play` to 0 when rate 0 is reached/crossed.
I think I can check filter and scratchprocess with valid and invalid input, but how to (wait and then) check if `play` was disabled and brakeActive is set to false?

**softStart()** gets arguments `deck`, `activate`, acceleration `factor`. From this values it also sets up a filter, enables play and initiates scratchprocess().
Same as with brake(): how to wait and check if softStartActive is false?

### CODE REVIEW

**Comment by illuusio** [comment]
Please resolve conflicts first so this can be looked at

**Comment by ronso0** [comment]
Thanks for looking at this, I commented out the empty test functions.
I think there's no way to use brake, spinBack or softStart from GUI, only with controller script, or mouse/hid script (didn't use that so far).
If you really need it in skin, I think there could be a workaround with mouse/hid-script.js and a custom buttom in any skin: connect a function in script to watch a certain skin CO and trigger script.brake() when that CO changes.

**Comment by ronso0** [comment]
Documentation is [here](http://www.mixxx.org/wiki/doku.php/midi_scripting#spinback_brake_and_soft_start_effect)

File: res/controllers/common-controller-scripts.js [hunk 1]
```diff
@@ -302,10 +302,14 @@ script.midiPitch = function (LSB, MSB, status) {
    Purpose: wrapper around engine.spinback() that can be directly mapped
             from xml for a spinback effect
             e.g: <key>script.spinback</key>
-   Input:   channel, control, value, status, group
+   Input:   channel, control, value, status, group, factor (optional), start rate (optional)
    Output:  none
    -------- ------------------------------------------------------ */
-script.spinback = function(channel, control, value, status, group) {
+script.spinback = function(channel, control, value, status, group, factor, rate) {
```
Comment (rryan) [inline res/controllers/common-controller-scripts.js]: rate is unused, remove it?

File: res/controllers/common-controller-scripts.js [hunk 2]
```diff
@@ -302,10 +302,14 @@ script.midiPitch = function (LSB, MSB, status) {
...
+    // set default factor
+    if (factor === undefined) {
+        factor = 1;
+    }
     // disable on note-off or zero value note/cc
     engine.spinback(parseInt(group.substring(8,9)), ((status & 0xF0) !== 0x80 && value > 0));
```
Comment (rryan) [inline res/controllers/common-controller-scripts.js]: did you mean to pass factor in as the 3rd argument?

File: res/controllers/common-controller-scripts.js (hunk 1 above)
Comment (rryan) [inline res/controllers/common-controller-scripts.js]: or maybe add it as the 4th argument to engine.spinback?

**Comment by ronso0** [comment]
I continue to work on this, [**WIP 1**] **fixing brake used with very low deceleration factors**, and address the comments.

As mentioned in the description, I have certain problems writing tests. Since I didn't receive any feedback on this I'll remove [**WIP 2**] **write tests** here so someone else can pick it up.

**Comment by Pegasus-RPG** [comment]
@ywwg Can you help with the test writing here?

**Comment by Pegasus-RPG** [comment]
( @uklotzde the Engine label is for the main mixing engine, not the controller one. :) )

**Comment by ronso0** [comment]
Brake/spinBack can be interrupted/stopped by scratching, but I just noticed it can't by stopping the deck.
I try to add a condition for this purpose.

**Comment by ronso0** [comment]
Everything is fixed, comments addressed and also the wiki is up-to-date now.
Ready to test!

**Comment by daschuer** [comment]
It would be really great to have this in the "engine" though, accessible from the skin as well.

**Comment by daschuer** [comment]
I do not see anything odd. Does anyone have controller ready to test it?

**Comment by Be-ing** [comment]
What's going on with this PR? Should we merge it to the 2.1 branch?

**Comment by ronso0** [comment]
IMO yes. There are no fundamental changes, it just improves brake and softStart. I tested any scenario coming to mind, incl. those two functions interrupting each other, like I did with the previous PR where I introduced softStart.
Just give it a try. Btw it's convenient to map both functions to play controls close to the shift button so you can trigger them with one hand.

**Comment by ronso0** [comment]
We shouldn't forget about the test though. As I mentioned earlier, I'm not capable of writing those, so someone with more experience needs to tackle this

**Comment by Be-ing** [comment]
Okay, then please open a bug report on Launchpad so we don't forget it.

**Pull request merged**